[DNM] sqlccl: create common partitioning tests

danhhz opened the pull request:
When I started in on repartitioning, it was obvious that we're duplicating a
LOT of code in the tests of various partitioning components. In the interest of
only having to add tests in one place, create a common set that can be used.

@benesch I had to stop here for the day, so it needs a bit more polish, but what do you think of the approach?

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20287)
<!-- Reviewable:end -->

danhhz committed:
sqlccl: refactor partitioning tests

They needed to be DRY'd. It was quickly becoming obvious that we were
going to start losing test coverage because adding a new edge case would
require finding all the various tests and updating them.

This introduces `partitioningTest` which has a shorthand for writing the
various relevant parts and a `parse` method which turns them into the
internal datastructures directly used by tests. It also adds an initial
method that returns a set of manually constructed interesting test cases
and well as a randomized test for each column type.

Release note: None

danhhz committed:
sqlccl: add RepartitioningFastPathAvailable

RepartitioningFastPathAvailable returns true when the schema change to
validate existing data can be skipped.

Each partitioned index guarantees that all rows in that index belong to
one of its partitions. Certain repartitionings (removing partitioning
entirely, a LIST partitioning that is a superset of the previous
partitioning, etc) can assume the the existing data meets this guarantee
without checking it via a sort of "inductive proof".

This method is added and tested, but not used yet.

Release note: None

danhhz commented:
General approach was validated. Closing this in favor of a new PR
